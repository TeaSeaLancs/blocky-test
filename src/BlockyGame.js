import React from 'react';
import BlockGrid from './BlockGrid';
import TransitionsGrid from './TransitionsGrid';

import useBlockGameState from './util/useBlockGameState';

// It seems a bit odd at first glance to have an initial value which we copy into internal component state
// but the benefit to this is that we can encapsulate all the game logic into this component and have the
// component just reset itself when it gets a different grid passed in.
export default function BlockyGame({ initialGrid }) {
  // A custom hook makes it easy to isolate all the game logic and have it hookable into a UI component
  // I mean realistically, it's only being used in one component so it's a bit overkill but it does
  // allow you to have nice, small components :)
  const [
    { grid, transitions },
    { removeBlock, clearTransitions },
  ] = useBlockGameState({ grid: initialGrid });

  // The BlockGrid and TransitionsGrid are separate so that we always have a component rendering the
  // grid as it's meant to be, and then a separate component can handle the niceties of displaying animations :)
  return (
    <div id="game">
      <BlockGrid grid={grid} onBlockRemove={removeBlock} />
      <TransitionsGrid
        grid={grid}
        transitions={transitions}
        onTransitionEnd={clearTransitions}
      />
    </div>
  );
}
