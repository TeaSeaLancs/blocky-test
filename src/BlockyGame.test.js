import React from 'react';
import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';

import BlockyGame from './BlockyGame';

import createSampleGrid from './__fixtures__/SampleGrid';

// We mock out the removeBlockAndUpdateGrid import here because we don't want to have to test its logic
// while we're testing this component.
jest.mock('./util/removeBlockAndUpdateGrid');

describe('BlockyGame', () => {
  let grid;

  beforeEach(() => {
    grid = createSampleGrid();
  });

  it('renders the game from the grid provided', () => {
    const { asFragment } = render(<BlockyGame initialGrid={grid} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
