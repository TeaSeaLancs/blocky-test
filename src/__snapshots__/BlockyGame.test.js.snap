// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockyGame renders the game from the grid provided 1`] = `
<DocumentFragment>
  <div
    id="game"
  >
    <div
      class="grid"
    >
      <div
        class="block active"
        data-testid="block-0-0"
        data-x="0"
        data-y="0"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-0-1"
        data-x="0"
        data-y="1"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-0-2"
        data-x="0"
        data-y="2"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-0-3"
        data-x="0"
        data-y="3"
        style="background-color: yellow;"
      />
      <div
        class="block active"
        data-testid="block-0-4"
        data-x="0"
        data-y="4"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-1-0"
        data-x="1"
        data-y="0"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-1-1"
        data-x="1"
        data-y="1"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-1-2"
        data-x="1"
        data-y="2"
        style="background-color: red;"
      />
      <div
        class="block active"
        data-testid="block-1-3"
        data-x="1"
        data-y="3"
        style="background-color: yellow;"
      />
      <div
        class="block active"
        data-testid="block-1-4"
        data-x="1"
        data-y="4"
        style="background-color: yellow;"
      />
      <div
        class="block active"
        data-testid="block-2-0"
        data-x="2"
        data-y="0"
        style="background-color: yellow;"
      />
      <div
        class="block active"
        data-testid="block-2-1"
        data-x="2"
        data-y="1"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-2-2"
        data-x="2"
        data-y="2"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-2-3"
        data-x="2"
        data-y="3"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-2-4"
        data-x="2"
        data-y="4"
        style="background-color: red;"
      />
      <div
        class="block active"
        data-testid="block-3-0"
        data-x="3"
        data-y="0"
        style="background-color: red;"
      />
      <div
        class="block active"
        data-testid="block-3-1"
        data-x="3"
        data-y="1"
        style="background-color: yellow;"
      />
      <div
        class="block active"
        data-testid="block-3-2"
        data-x="3"
        data-y="2"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-3-3"
        data-x="3"
        data-y="3"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-3-4"
        data-x="3"
        data-y="4"
        style="background-color: red;"
      />
      <div
        class="block active"
        data-testid="block-4-0"
        data-x="4"
        data-y="0"
        style="background-color: green;"
      />
      <div
        class="block active"
        data-testid="block-4-1"
        data-x="4"
        data-y="1"
        style="background-color: red;"
      />
      <div
        class="block active"
        data-testid="block-4-2"
        data-x="4"
        data-y="2"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-4-3"
        data-x="4"
        data-y="3"
        style="background-color: blue;"
      />
      <div
        class="block active"
        data-testid="block-4-4"
        data-x="4"
        data-y="4"
        style="background-color: blue;"
      />
    </div>
    <div
      class="grid floating"
    />
  </div>
</DocumentFragment>
`;
